File: SuperUglyNumber/SuperUglyNumber.py
class Solution(object):
    def nthSuperUglyNumber(self, n, primes):
        """
        :type n: int
        :type primes: List[int]
        :rtype: int
        """
        index = [0] * len(primes)
        ugly = [1]
        for i in xrange(n):
        	nextUgly = primes[0] * ugly[index[0]]
        	for j in xrange(len(primes)):
        		nextUgly = min(nextUgly, primes[j] * ugly[index[j]])
        	for j in xrange(len(primes)):
        		if nextUgly == primes[j] * ugly[index[j]]:
        			index[j] += 1
        	ugly.append(nextUgly)
        return ugly[n - 1]

File: SuperUglyNumber/SuperUglyNumber.cpp
class Solution {
public:
    int nthSuperUglyNumber(int n, vector<int>& primes) {
        vector<int> index(primes.size(), 0), ugly;
        ugly.push_back(1);
        for(int i = 1; i < n;i++) {
        	int next = ugly[index[0]] * primes[0];
        	for(int j = 0;j < primes.size(); j++)
        		next = min(next, ugly[index[j]] * primes[j]);
        	for(int j = 0;j < primes.size(); j++)
        		if(next == ugly[index[j]] * primes[j])
        			index[j]++;
        	ugly.push_back(next);
        }
        return ugly[n - 1];
    }
};